// playwright-bdd/dist/features/load.js
"use strict";
/**
 * Load features.
 *
 * See: https://github.com/cucumber/cucumber-js/blob/main/src/api/load_sources.ts
 * See: https://github.com/cucumber/cucumber-js/blob/main/src/api/gherkin.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeaturesLoader = exports.resolveFeatureFiles = void 0;
const gherkin_streams_1 = require("@cucumber/gherkin-streams");
const gherkin_utils_1 = require("@cucumber/gherkin-utils");
const paths_1 = require("../utils/paths");
const utils_1 = require("../utils");
function resolveFeatureFiles(cwd, patterns) {
    return (0, paths_1.resolveFiles)(cwd, (0, utils_1.toArray)(patterns), 'feature');
}
exports.resolveFeatureFiles = resolveFeatureFiles;
class FeaturesLoader {
    constructor() {
        this.gherkinQuery = new gherkin_utils_1.Query();
        this.parseErrors = [];
    }
    /**
     * Loads and parses feature files.
     * - featureFiles should be absolute.
     *   See: https://github.com/cucumber/gherkin-streams/blob/main/src/GherkinStreams.ts#L36
     * - if options.relativeTo is provided, uri in gherkin documents will be relative to it.
     *   See: https://github.com/cucumber/gherkin-streams/blob/main/src/SourceMessageStream.ts#L31
     * - options.defaultDialect is 'en' by default.
     *   See: https://github.com/cucumber/gherkin-streams/blob/main/src/makeGherkinOptions.ts#L5
     */
    async load(featureFiles, options) {
        this.gherkinQuery = new gherkin_utils_1.Query();
        this.parseErrors = [];
        // Without this early return gherkinFromPaths() produced weird behavior
        // for reporters: it does not keep exit code
        // See: https://github.com/vitalets/playwright-bdd/issues/200
        if (!featureFiles.length)
            return;
        await gherkinFromPaths(featureFiles, options, (envelope) => {
            this.gherkinQuery.update(envelope);
            if (envelope.parseError) {
                this.parseErrors.push(envelope.parseError);
            }
        });
    }
    getDocumentsCount() {
        return this.gherkinQuery.getGherkinDocuments().length;
    }
    getDocumentsWithPickles() {
        return this.gherkinQuery.getGherkinDocuments().map((gherkinDocument) => {
            const pickles = this.getDocumentPickles(gherkinDocument);
            return { ...gherkinDocument, pickles };
        });
    }
    getDocumentPickles(gherkinDocument) {
        return this.gherkinQuery
            .getPickles()
            .filter((pickle) => gherkinDocument.uri === pickle.uri)
            .map((pickle) => this.getPickleWithLocation(pickle));
    }
    getPickleWithLocation(pickle) {
        const lastAstNodeId = pickle.astNodeIds[pickle.astNodeIds.length - 1];
        const location = this.gherkinQuery.getLocation(lastAstNodeId);
        return { ...pickle, location };
    }
}
exports.FeaturesLoader = FeaturesLoader;
async function gherkinFromPaths(paths, options, onEnvelope) {
    return new Promise((resolve, reject) => {
        const gherkinMessageStream = gherkin_streams_1.GherkinStreams.fromPaths(paths, options);
        gherkinMessageStream.on('data', onEnvelope);
        gherkinMessageStream.on('end', resolve);
        gherkinMessageStream.on('error', reject);
    });
}
